### ISSUE TITLE
Define behavior on multiple matching TLS roles

### ISSUE THREAD

**Issue by bluecmd** [hashicorp/vault#5923]
**Is your feature request related to a problem? Please describe.**
As [discussed](https://groups.google.com/d/msg/vault-tool/CZvdluilxpM/vNa57IywBQAJ) in the scenario where you have a setup like below, the behavior is undefined.

- Role 1: policy=server, certificate=@ca.crt
- Role 2: policy=server-master, certificate=@ca.crt, allowed_common_names=["master1.domain.tld", "master2.domain.tld"]

I.e. one role matching any certificate and one matching some specific CNs. For the server master1.domain.tld Vault will according to the documentation pick a role, but how that role is chosen is not documented, and I believe it depends on the backend currently (possibly non-deterministic?).

Documentation says:
```
name (string: "") - Authenticate against only the named certificate role, returning its policy list if successful. If not set, defaults to trying all certificate roles and **returning any one that matches.**
```

The problem is that `any one` is not defined in the case of multiple matches.

**Describe the solution you'd like**
The order to be defined regardless of backend. For example taking the lexicographically first role.
That way we can reason about what would be the default role and use the default (empty) role name in the client authentication request for every server.

**Describe alternatives you've considered**
Currently we have hard-coded all the roles into the specific servers which is messy and makes refactors hard to do.

**Explain any additional use-cases**
This would make troubleshooting why certain policies are in effect easier.

**Additional context**
[Discussion on the mailing list](https://groups.google.com/d/msg/vault-tool/CZvdluilxpM/vNa57IywBQAJ)

**Comment by jefferai** [comment]
Arguably keeping it undefined is correct behavior. If it can't be determined specifically which role should match, you shouldn't depend on the result. (Go does this for map iteration for exactly that reason.)

**Comment by bluecmd** [comment]
I strongly disagree. There is a huge difference between limitations in a data structure and the guarantees on an API.

If you keep it undefined I ask you to document it as such with a warning, or consider erroring out instead if there are multiple matches. That's not what I want, but I'm pretty sure you'll find more people in the community supportive of their authentication system's behavior being strict rather than undefined.

**Comment by jefferai** [comment]
The problem is there isn't any obvious ordering that should be applied. Lexograpgical by role? Lexograpgical by CA CN? Issue time? Expiration time? Serial number? Matching Locality?

To be clear I think the right solution is to match against all and get the union of policies. I just don't think there's any obviously correct ordering if there's a limit to one, and if there's no obviously right order it should either not try to impose one or, maybe, error out. (There is a client side fix after all.)

**Comment by bluecmd** [comment]
I agree that there are multiple orders and no "correct" one, but *any* order would make it possible for me to reason about it - compared to now. Changing to a order would make the system better - not perfect, but better, so why not?

As for the client side fix, yes there is - but knowing when to use it requires perfect information.
It worked for me in this case because I'm both the administrator of the roles as well as the client developer, but that's obviously rarely the case. If you're not 100% aware of the roles that are present on the backend now and in the future you **have** to specify the role or you risk undefined behavior - which should be unacceptable for anyone doing authentication.

**Comment by jefferai** [comment]
> Changing to a order would make the system better - not perfect, but better, so why not?

I disagree though. It's patching around something that can already be managed via different mounts on the server or more specific client behavior, and worse...

> It also cannot break backwards-compatibility as the behavior is undefined to day.

...it will, if it's dependent on storage backend ordering. What won't break your backwards compat may well break someone else's.

> The union of all policies would indeed be great but from the mailinglist discussions and reading the PRs that created this feature it seems like that ship has sailed 

I'd have to revisit the discussion but all ships can come back to port.

**Comment by bluecmd** [comment]
> ...it will, if it's dependent on storage backend ordering. What won't break your backwards compat may well break someone else's.

The reason I'm bringing this up in the first place is that we used to be dependent on the ordering. We've been running with multiple roles for years, but then we upgraded Vault and that changed the order the roles are iterated. And here I am, trying to solve it once and for all.

Maybe I'm misunderstanding, but you're talking about somebody depending on undefined behavior as a supported use case that should not be disrupted? That seems odd to me. The reason one defines something as undefined behavior is to discourage people from using it and allowing changing the internals in the future, yet you seem to argue that it shouldn't be changed.

**Comment by bluecmd** [comment]
@hsimon-hashicorp It's pretty easy to check the documentation if this is still an issue. You don't need me in order to do that.

**Comment by hsimon-hashicorp** [comment]
> @hsimon-hashicorp It's pretty easy to check the documentation if this is still an issue. You don't need me in order to do that.

Hello! Thanks for your reply. Due to the nature of the subsequent discussion and the age of the issue, I was not clear on what the request was. For that, I apologize. Additionally, functionality can change while documentation is not always updated in cadence, although ideally it should be.

@schavis and @yhyakuna the issue seems to be that the documentation for TLS Certificates in Auth Methods is incorrect for the second sentence under the ["name" parameter](https://developer.hashicorp.com/vault/api-docs/auth/cert#name-6).

Thanks again,
Heather
